homeassistant/components/logbook/helpers.py

```diff
@@ -0,0 +1,190 @@
+    ]
+
+
+def async_determine_event_types(
+    hass: HomeAssistant, entity_ids: list[str] | None, device_ids: list[str] | None
+) -> tuple[str, ...]:
+    """Reduce the event types based on the entity ids and device ids."""
+    external_events: dict[
+        str, tuple[str, Callable[[LazyEventPartialState], dict[str, Any]]]
+    ] = hass.data.get(DOMAIN, {})
+    if not entity_ids and not device_ids:
+        return (*ALL_EVENT_TYPES_EXCEPT_STATE_CHANGED, *external_events)
+    config_entry_ids: set[str] = set()
+    intrested_event_types: set[str] = set()
+
+    if entity_ids:
+        #
+        # Home Assistant doesn't allow firing events from
+        # entities so we have a limited list to check
+        #
+        # automations and scripts can refer to entities
+        # but they do not have a config entry so we need
+        # to add them.
+        #
+        # We also allow entity_ids to be recorded via
+        # manual logbook entries.
+        #
+        intrested_event_types |= ENTITY_EVENTS_WITHOUT_CONFIG_ENTRY
+
+    if device_ids:
+        dev_reg = dr.async_get(hass)
+        for device_id in device_ids:
+            if (device := dev_reg.async_get(device_id)) and device.config_entries:
+                config_entry_ids |= device.config_entries
+        interested_domains: set[str] = set()
+        for entry_id in config_entry_ids:
+            if entry := hass.config_entries.async_get_entry(entry_id):
+                interested_domains.add(entry.domain)
+        for external_event, domain_call in external_events.items():
+            if domain_call[0] in interested_domains:
+                intrested_event_types.add(external_event)
+
+    return tuple(
+        event_type
+        for event_type in (EVENT_LOGBOOK_ENTRY, *external_events)
+        if event_type in intrested_event_types
+    )
+
+
+@callback
+def async_subscribe_events(
+    hass: HomeAssistant,
+    subscriptions: list[CALLBACK_TYPE],
+    target: Callable[[Event], None],
+    event_types: tuple[str, ...],
+    entity_ids: list[str] | None,
+    device_ids: list[str] | None,
+) -> None:
+    """Subscribe to events for the entities and devices or all.
+
+    These are the events we need to listen for to do
+    the live logbook stream.
+    """
+    ent_reg = er.async_get(hass)
+    if entity_ids or device_ids:
+        entity_ids_set = set(entity_ids) if entity_ids else set()
+        device_ids_set = set(device_ids) if device_ids else set()
+
+        @callback
+        def _forward_events_filtered(event: Event) -> None:
+            event_data = event.data
+            if (
+                entity_ids_set and event_data.get(ATTR_ENTITY_ID) in entity_ids_set
+            ) or (device_ids_set and event_data.get(ATTR_DEVICE_ID) in device_ids_set):
+                target(event)
+
+        event_forwarder = _forward_events_filtered
+    else:
+
+        @callback
+        def _forward_events(event: Event) -> None:
```

Comment:
Why even wrap it? We can test it's labeled a callback and be done with it?

Reply:
It was more complex a few iterations ago.  I'll clean it up some more

Reply:
c281ad160913381fe6bcbfef591032a23398db19